File: 8vim/src/main/java/inc/flide/vim8/keyboardactionlisteners/ButtonKeypadActionListener.java
package inc.flide.vim8.keyboardactionlisteners;

import android.view.View;
import com.hijamoya.keyboardview.KeyboardView;
import inc.flide.vim8.MainInputMethodService;
import inc.flide.vim8.models.CustomKeycode;

public class ButtonKeypadActionListener extends KeypadActionListener implements KeyboardView.OnKeyboardActionListener {
    private static final int NOT_A_KEY = -1;

    public ButtonKeypadActionListener(MainInputMethodService mainInputMethodService, View view) {
        super(mainInputMethodService, view);
    }

    @Override
    public void onKey(int primaryCode, int[] keyCodes) {
        if (primaryCode == NOT_A_KEY) {
            return;
        }
        if (primaryCode == CustomKeycode.SHIFT_TOGGLE.keyCode) {
            handleShiftToggle();
        } else if (primaryCode == CustomKeycode.CTRL_TOGGLE.keyCode) {
            mainInputMethodService.performCtrlToggle();
        } else {
            super.handleInputKey(primaryCode, 0);
        }
    }

    private void handleShiftToggle() {
        if (mainInputMethodService.getShiftState() == MainInputMethodService.State.OFF) {
            mainInputMethodService.performShiftToggle();
        } else {
            mainInputMethodService.setShiftState(MainInputMethodService.State.OFF);
        }
    }

    @Override
    public void swipeLeft() {
    }

    @Override
    public void swipeRight() {
    }

    @Override
    public void swipeDown() {
        mainInputMethodService.hideKeyboard();
    }

    @Override
    public void swipeUp() {
    }

    @Override
    public void onPress(int primaryCode) {
    }

    @Override
    public void onRelease(int primaryCode) {
    }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/ime/actionlisteners/ButtonKeypadActionListener.kt
package inc.flide.vim8.ime.actionlisteners

import android.view.View
import com.hijamoya.keyboardview.KeyboardView
import inc.flide.vim8.Vim8ImeService
import inc.flide.vim8.ime.layout.models.CustomKeycode

private const val NOT_A_KEY = -1

class ButtonKeypadActionListener(vim8ImeService: Vim8ImeService, view: View) :
    KeypadActionListener(vim8ImeService, view), KeyboardView.OnKeyboardActionListener {
    override fun onKey(primaryCode: Int, keyCodes: IntArray) {
        when (primaryCode) {
            CustomKeycode.SHIFT_TOGGLE.keyCode -> handleShiftToggle()
            NOT_A_KEY -> {}
            else -> super.handleInputKey(primaryCode, 0)
        }
    }

    private fun handleShiftToggle() {
        if (vim8ImeService.shiftState == Vim8ImeService.State.OFF) {
            vim8ImeService.performShiftToggle()
        } else {
            vim8ImeService.shiftState = Vim8ImeService.State.OFF
        }
    }

    override fun swipeLeft() {}
    override fun swipeRight() {}
    override fun swipeDown() {
        vim8ImeService.hideKeyboard()
    }

    override fun swipeUp() {}
    override fun onPress(primaryCode: Int) {}
    override fun onRelease(primaryCode: Int) {}
}
